firedrake/formmanipulation.py

```diff
@@ -98,76 +118,42 @@ def argument(self, o):
-        if isinstance(indices, int):
-            indices = (indices, )
+        W = subspace(V, indices)
+        a = Argument(W, o.number(), part=o.part())
+        a = (a, ) if len(W) == 1 else split(a)
 
-        if len(indices) == 1:
-            W = V_is[indices[0]]
-            W = FunctionSpace(W.mesh(), W.ufl_element())
-            a = (Argument(W, o.number(), part=o.part()), )
-        else:
-            W = MixedFunctionSpace([V_is[i] for i in indices])
-            a = split(Argument(W, o.number(), part=o.part()))
         args = []
-        for i in range(len(V_is)):
+        for i in range(len(V)):
             if i in indices:
                 c = indices.index(i)
                 a_ = a[c]
                 if len(a_.ufl_shape) == 0:
-                    args += [a_]
+                    args.append(a_)
                 else:
-                    args += [a_[j] for j in numpy.ndindex(a_.ufl_shape)]
+                    args.extend(a_[j] for j in numpy.ndindex(a_.ufl_shape))
             else:
-                args += [Zero()
-                         for j in numpy.ndindex(V_is[i].value_shape)]
+                args.extend(Zero() for j in numpy.ndindex(V[i].value_shape))
         return self._arg_cache.setdefault(o, as_vector(args))
 
     def cofunction(self, o):
         V = o.function_space()
 
-        # Not on a mixed space, just return ourselves.
         if len(V) == 1:
+            # Not on a mixed space, just return ourselves.
             return o
 
-        # We only need the test space for Cofunction
+        # We only need the test space for Cofunction		￼
         indices = self.blocks[0]
-        V_is = V.subfunctions
-
-        # Only one index provided.
-        if isinstance(indices, int):
-            indices = (indices, )
-
-        # for two-forms, the cofunction should only
-        # be returned for the diagonal blocks, so
-        # if we are asked for an off-diagonal block
-        # then we return a zero form, analogously to
-        # the off components of arguments.
-        if len(self.blocks) == 2:
-            itest, itrial = self.blocks
-            on_diag = (itest == itrial)
-        else:
-            on_diag = True
-
-        # if we are on the diagonal, then return a Cofunction
-        # in the relevant subspace that points to the data in
-        # the full space. This means that the right hand side
-        # of the fieldsplit problem will be correct.
-        if on_diag:
-            if len(indices) == 1:
-                i = indices[0]
-                W = V_is[i]
-                W = DualSpace(W.mesh(), W.ufl_element())
-                c = Cofunction(W, val=o.subfunctions[i].dat)
-            else:
-                W = MixedFunctionSpace([V_is[i] for i in indices])
-                c = Cofunction(W, val=MixedDat(o.dat[i] for i in indices))
+        if len(indices) == 1:
+            i = indices[0]
+            W = V[i]
+            W = DualSpace(W.mesh(), W.ufl_element())
```

Comment:
Can you use the `subspace` function here too? It will need a switch adding for primal/dual but I think it would be nicer to use `subspace` for both.

I think it would also be nice to lift `subspace` out, possibly into one of the functionspace files. It's useful more generally - see for example line 678 in `slate.py` where we have the same if/else block to construct the subspace.

Reply:
I was thinking to extend `FunctionSpace.__getitem__` to allow a list index return the subspace, since this is how sub-arrays are extracted in numpy.

Reply:
I was actually thinking exactly the same thing the other day.

It would be great to have it for `Function` too so that you can get a `Mixed` subfunction viewing the relevant subcomponents. `ExtractSubBlock.cofunction` would then basically just be calling `cofunction[idxs]`.
There are definitely other places where that would be useful too, for example in the `EnsembleFunction` I've been working on [here](https://github.com/firedrakeproject/firedrake/blob/dfe98bf906b0c063822abec533fccc1d3bea67f3/firedrake/ensemblefunction.py#L59).

What does `FunctionSpace.collapse()` do? It just looks like it would return a copy of the `FunctionSpace`.

Reply:
@ksagiyam what do you think about this __getitem__ idea? Currenty, indexing by a scalar returns `IndexedProxyFunctionSpace`, but indexing by a list of a single item would return the collapsed version of this space.

Reply:
`__getitem__` for `Function` is implemented in UFL so the changes would be needed there I think.

Reply:
I'm not to worried for the `Function` case, that one seems more involved, as they are Expressions in the UFL sense, and extracting a component naturally means a scalar compoment, but here it seems that we want a MixedFunctionSpace (possibly vector-valued) component.

Reply:
It could also be added to `sub` or `subfunctions`, although they both also have issues.
`sub` explicitly says it is for bcs, and `subfunctions` is already a tuple not a method so it would probably have to become a local class with `__getitem__` overridden so we could deal with single or multiple indices.

Reply:
That sounds slightly odd to me. Can you instead rewrite `collapse()` so that it would take a list of indices as an optional argument?

@JHopeCollins `collapse()` basically forgets parents. They do something nontrivial in ProxyFunctionSpace and in MixedFunctionSpace.

Reply:
Ok thanks.

It does strip the parent information, I don't think the name `collapse` would be where I would look for this functionality - it would be more intuitive in subfunctions or sub.

Reply:
I'll go back to the old `subspace` method I had for splitting purposes